File: java/src/main/java/dojo/supermarket/model/ShoppingCart.java
package dojo.supermarket.model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ShoppingCart {

    // Constants for magic numbers
    private static final double DEFAULT_ITEM_QUANTITY = 1.0;
    private static final int TWO_FOR_AMOUNT_QUANTITY = 2;
    private static final int THREE_FOR_TWO_QUANTITY = 3;
    private static final int FIVE_FOR_AMOUNT_QUANTITY = 5;
    private static final double PERCENTAGE_DIVISOR = 100.0;

    // Constants for discount description strings
    private static final String THREE_FOR_TWO_DESCRIPTION = "3 for 2";
    private static final String TWO_FOR_PREFIX = "2 for ";
    private static final String FOR_SEPARATOR = " for ";
    private static final String PERCENT_OFF_SUFFIX = "% off";

    private final List<ProductQuantity> items = new ArrayList<>();
    private final Map<Product, Double> productQuantities = new HashMap<>();

    List<ProductQuantity> getItems() {
        return Collections.unmodifiableList(items);
    }

    void addItem(Product product) {
        addItemQuantity(product, DEFAULT_ITEM_QUANTITY);
    }

    Map<Product, Double> productQuantities() {
        return Collections.unmodifiableMap(productQuantities);
    }

    public void addItemQuantity(Product product, double quantity) {
        items.add(new ProductQuantity(product, quantity));
        if (productQuantities.containsKey(product)) {
            productQuantities.put(product, productQuantities.get(product) + quantity);
        } else {
            productQuantities.put(product, quantity);
        }
    }

    void handleOffers(Receipt receipt, Map<Product, Offer> offers, SupermarketCatalog catalog) {
        for (Product p: productQuantities().keySet()) {
            double quantity = productQuantities.get(p);
            if (offers.containsKey(p)) {
                Offer offer = offers.get(p);
                double unitPrice = catalog.getUnitPrice(p);
                int quantityAsInt = (int) quantity;
                Discount discount = null;
                int x = 1;
                if (offer.offerType == SpecialOfferType.THREE_FOR_TWO) {
                    x = THREE_FOR_TWO_QUANTITY;

                } else if (offer.offerType == SpecialOfferType.TWO_FOR_AMOUNT) {
                    x = TWO_FOR_AMOUNT_QUANTITY;
                    if (quantityAsInt >= TWO_FOR_AMOUNT_QUANTITY) {
                        double total = offer.argument * (quantityAsInt / x) + quantityAsInt % TWO_FOR_AMOUNT_QUANTITY * unitPrice;
                        double discountN = unitPrice * quantity - total;
                        discount = new Discount(p, TWO_FOR_PREFIX + offer.argument, -discountN);
                    }

                } if (offer.offerType == SpecialOfferType.FIVE_FOR_AMOUNT) {
                    x = FIVE_FOR_AMOUNT_QUANTITY;
                }
                int numberOfXs = quantityAsInt / x;
                if (offer.offerType == SpecialOfferType.THREE_FOR_TWO && quantityAsInt > TWO_FOR_AMOUNT_QUANTITY) {
                    double discountAmount = quantity * unitPrice - ((numberOfXs * TWO_FOR_AMOUNT_QUANTITY * unitPrice) + quantityAsInt % THREE_FOR_TWO_QUANTITY * unitPrice);
                    discount = new Discount(p, THREE_FOR_TWO_DESCRIPTION, -discountAmount);
                }
                if (offer.offerType == SpecialOfferType.TEN_PERCENT_DISCOUNT) {
                    discount = new Discount(p, offer.argument + PERCENT_OFF_SUFFIX, -quantity * unitPrice * offer.argument / PERCENTAGE_DIVISOR);
                }
                if (offer.offerType == SpecialOfferType.FIVE_FOR_AMOUNT && quantityAsInt >= FIVE_FOR_AMOUNT_QUANTITY) {
                    double discountTotal = unitPrice * quantity - (offer.argument * numberOfXs + quantityAsInt % FIVE_FOR_AMOUNT_QUANTITY * unitPrice);
                    discount = new Discount(p, x + FOR_SEPARATOR + offer.argument, -discountTotal);
                }
                if (discount != null)
                    receipt.addDiscount(discount);
            }
        }
    }
}

File: ruby/lib/models/shopping_cart.rb
# Constants for magic numbers
DEFAULT_ITEM_QUANTITY = 1.0
TWO_FOR_AMOUNT_QUANTITY = 2
THREE_FOR_TWO_QUANTITY = 3
FIVE_FOR_AMOUNT_QUANTITY = 5
PERCENTAGE_DIVISOR = 100.0

# Constants for discount description strings
THREE_FOR_TWO_DESCRIPTION = "3 for 2"
TWO_FOR_PREFIX = "2 for "
FOR_SEPARATOR = " for "
PERCENT_OFF_SUFFIX = "% off"

class ShoppingCart

  def initialize
    @items = []
    @product_quantities = {}
  end

  def items
    Array.new @items
  end

  def add_item(product)
    add_item_quantity(product, DEFAULT_ITEM_QUANTITY)
    nil
  end

  def product_quantities
    @product_quantities
  end

  def add_item_quantity(product, quantity)
    @items << ProductQuantity.new(product, quantity)
    if @product_quantities.key?(product)
      product_quantities[product] = product_quantities[product] + quantity
    else
      product_quantities[product] = quantity
    end
  end

  def handle_offers(receipt, offers, catalog)
    for p in @product_quantities.keys do
      quantity = @product_quantities[p]
      if offers.key?(p)
        offer = offers[p]
        unit_price = catalog.unit_price(p)
        quantity_as_int = quantity.to_i
        discount = nil
        x = 1
        if offer.offer_type == SpecialOfferType::THREE_FOR_TWO
          x = THREE_FOR_TWO_QUANTITY

        elsif offer.offer_type == SpecialOfferType::TWO_FOR_AMOUNT
          x = TWO_FOR_AMOUNT_QUANTITY
          if quantity_as_int >= TWO_FOR_AMOUNT_QUANTITY
            total = offer.argument * (quantity_as_int / x) + quantity_as_int % TWO_FOR_AMOUNT_QUANTITY * unit_price
            discount_n = unit_price * quantity - total
            discount = Discount.new(p, TWO_FOR_PREFIX + offer.argument.to_s, discount_n)
          end

        end
        if offer.offer_type == SpecialOfferType:: FIVE_FOR_AMOUNT
          x = FIVE_FOR_AMOUNT_QUANTITY
        end
        number_of_x = quantity_as_int / x
        if offer.offer_type == SpecialOfferType::THREE_FOR_TWO && quantity_as_int > TWO_FOR_AMOUNT_QUANTITY
          discount_amount = quantity * unit_price - ((number_of_x * TWO_FOR_AMOUNT_QUANTITY * unit_price) + quantity_as_int % THREE_FOR_TWO_QUANTITY * unit_price)
          discount = Discount.new(p, THREE_FOR_TWO_DESCRIPTION, discount_amount)
        end
        if offer.offer_type == SpecialOfferType::TEN_PERCENT_DISCOUNT
          discount = Discount.new(p, offer.argument.to_s + PERCENT_OFF_SUFFIX, quantity * unit_price * offer.argument / PERCENTAGE_DIVISOR)
        end
        if offer.offer_type == SpecialOfferType::FIVE_FOR_AMOUNT && quantity_as_int >= FIVE_FOR_AMOUNT_QUANTITY
          discount_total = unit_price * quantity - (offer.argument * number_of_x + quantity_as_int % FIVE_FOR_AMOUNT_QUANTITY * unit_price)
          discount = Discount.new(p, x.to_s + FOR_SEPARATOR + offer.argument.to_s, discount_total)
        end

        receipt.add_discount(discount) if discount
      end
    end
  end

end
